--- a/fs/erofs/decompressor.c
+++ b/fs/erofs/decompressor.c
@@ -14,7 +14,21 @@
 
 #define LZ4_MAX_DISTANCE_PAGES	(DIV_ROUND_UP(LZ4_DISTANCE_MAX, PAGE_SIZE) + 1)
 #ifndef LZ4_DECOMPRESS_INPLACE_MARGIN
-#define LZ4_DECOMPRESS_INPLACE_MARGIN(srcsize)  (((srcsize) >> 8) + 32)
+static bool z_erofs_lz4_inplace_margin65;
+
+static int __init z_erofs_get_lz4_decompress_inplace_margin(char *str)
+{
+	if (!strcmp(str, "1"))
+		z_erofs_lz4_inplace_margin65 = true;
+
+	pr_info("erofs lz4 decompress inplace margin: %d\n",
+		z_erofs_lz4_inplace_margin65 ? 65 : 32);
+	return 0;
+}
+early_param("lz4asm.support", z_erofs_get_lz4_decompress_inplace_margin);
+
+#define LZ4_DECOMPRESS_INPLACE_MARGIN(srcsize) \
+	(((srcsize) >> 8) + (z_erofs_lz4_inplace_margin65 ? 65 : 32))
 #endif
 
 struct z_erofs_lz4_decompress_ctx {
